File: src/CircleAI.Speech.Cloud/DeepgramSpeechRecognizer.cs
// DeepgramSpeechRecognizer.cs
//
// (3.3.0) ISpeechRecognizer backed by Deepgram's /v1/listen endpoint.
// Single-shot HTTP POST with raw PCM (encoding=linear16) — same path
// the Patter Python SDK uses.

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace CircleAI.Speech.Cloud;

/// <summary>(3.3.0) Deepgram-backed <see cref="ISpeechRecognizer"/>.</summary>
public sealed class DeepgramSpeechRecognizer : ISpeechRecognizer
{
    private readonly HttpClient _http;
    private readonly DeepgramOptions _options;
    private readonly ILogger _logger;

    public DeepgramSpeechRecognizer(HttpClient http, DeepgramOptions options, ILogger<DeepgramSpeechRecognizer>? logger = null)
    {
        _http    = http    ?? throw new ArgumentNullException(nameof(http));
        _options = options ?? throw new ArgumentNullException(nameof(options));
        _logger  = (ILogger?)logger ?? NullLogger.Instance;

        if (_http.BaseAddress is null) _http.BaseAddress = options.BaseAddress;
    }

    public string BackendId    => "deepgram";
    public bool   IsConfigured => !string.IsNullOrWhiteSpace(_options.ApiKey);

    public async ValueTask<TranscriptionResult> TranscribeAsync(
        ReadOnlyMemory<byte> audioPcm16Mono,
        int                  sampleRateHz,
        string?              languageHint = null,
        CancellationToken    ct           = default)
    {
        if (!IsConfigured) return Empty();

        var path = $"/v1/listen?model={Uri.EscapeDataString(_options.Model)}&encoding=linear16&sample_rate={sampleRateHz}&channels=1&punctuate=true";
        if (!string.IsNullOrWhiteSpace(languageHint))
        {
            path += $"&language={Uri.EscapeDataString(languageHint)}";
        }

        using var msg = new HttpRequestMessage(HttpMethod.Post, path)
        {
            Content = new ByteArrayContent(audioPcm16Mono.ToArray()),
        };
        msg.Content!.Headers.ContentType = new MediaTypeHeaderValue("audio/raw");
        msg.Headers.Authorization = new AuthenticationHeaderValue("Token", _options.ApiKey);

        using var resp = await _http.SendAsync(msg, ct).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            _logger.LogWarning("Deepgram returned {Status}", resp.StatusCode);
            return Empty();
        }

        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);

        // Response shape: results.channels[0].alternatives[0].transcript
        if (!doc.RootElement.TryGetProperty("results", out var results)) return Empty();
        if (!results.TryGetProperty("channels", out var channels) || channels.GetArrayLength() == 0) return Empty();
        var firstChannel = channels[0];
        if (!firstChannel.TryGetProperty("alternatives", out var alts) || alts.GetArrayLength() == 0) return Empty();
        var firstAlt = alts[0];

        var text = firstAlt.TryGetProperty("transcript", out var t) ? t.GetString() ?? "" : "";

        var segments = new List<TranscribedSegment>();
        if (firstAlt.TryGetProperty("words", out var words) && words.ValueKind == JsonValueKind.Array)
        {
            foreach (var w in words.EnumerateArray())
            {
                segments.Add(new TranscribedSegment(
                    Text:       w.TryGetProperty("word",       out var ww) ? ww.GetString() ?? "" : "",
                    Offset:     TimeSpan.FromSeconds(w.TryGetProperty("start",      out var ss) ? ss.GetDouble() : 0d),
                    Duration:   TimeSpan.FromSeconds(w.TryGetProperty("end",        out var ee) ? ee.GetDouble() - (w.TryGetProperty("start", out var s2) ? s2.GetDouble() : 0d) : 0d),
                    Language:   languageHint,
                    Confidence: w.TryGetProperty("confidence", out var cc) ? (float)cc.GetDouble() : 0f));
            }
        }

        var duration = doc.RootElement.TryGetProperty("metadata", out var meta) && meta.TryGetProperty("duration", out var d)
            ? TimeSpan.FromSeconds(d.GetDouble())
            : TimeSpan.Zero;

        return new TranscriptionResult(text, languageHint, segments, duration);
    }

    private static TranscriptionResult Empty() =>
        new(string.Empty, null, Array.Empty<TranscribedSegment>(), TimeSpan.Zero);
}

File: python/src/circle_ai/speech/cloud/deepgram_speech_recognizer.py
# speech/cloud/deepgram_speech_recognizer.py
#
# Port of CircleAI.Speech.Cloud/DeepgramSpeechRecognizer.cs (C# — the EXACT spec).
#
# (3.3.0) ISpeechRecognizer backed by Deepgram's /v1/listen endpoint. Single-shot
# HTTP POST with raw PCM (encoding=linear16), "Token <key>" auth, and the
# results.channels[0].alternatives[0] response shape (transcript + per-word
# segments). Fail-soft on missing key / non-2xx.
#
# The C# drives HttpClient directly; the Python port injects the shared
# circle_ai.integration.http.IHttpFetcher. The raw PCM body rides body_bytes with
# the audio/raw content-type; JSON is read off resp.json().

from __future__ import annotations

import logging
from datetime import timedelta
from typing import List, Optional

from ...integration.http import HttpRequest, IHttpFetcher
from ..contracts import ISpeechRecognizer, TranscribedSegment, TranscriptionResult
from ._audio_http import combine_uri, is_null_or_whitespace
from .options import DeepgramOptions

_logger = logging.getLogger("CircleAI.Speech.Cloud.DeepgramSpeechRecognizer")


def _empty() -> TranscriptionResult:
    return TranscriptionResult("", None, (), timedelta(0))


def _escape(value: str) -> str:
    from urllib.parse import quote

    return quote(value, safe="-_.~")


class DeepgramSpeechRecognizer(ISpeechRecognizer):
    """(3.3.0) Deepgram-backed :class:`ISpeechRecognizer`.

    Mirrors ``CircleAI.Speech.Cloud.DeepgramSpeechRecognizer``.
    """

    def __init__(
        self,
        http: IHttpFetcher,
        options: DeepgramOptions,
        logger: Optional[logging.Logger] = None,
    ) -> None:
        if http is None:
            raise ValueError("http must not be None")
        if options is None:
            raise ValueError("options must not be None")
        self._http = http
        self._options = options
        self._logger = logger if logger is not None else _logger

    @property
    def backend_id(self) -> str:
        return "deepgram"

    @property
    def is_configured(self) -> bool:
        return not is_null_or_whitespace(self._options.api_key)

    async def transcribe_async(
        self,
        audio_pcm16_mono: bytes,
        sample_rate_hz: int,
        language_hint: Optional[str] = None,
        ct: object = None,
    ) -> TranscriptionResult:
        if not self.is_configured:
            return _empty()

        path = (
            f"/v1/listen?model={_escape(self._options.model)}&encoding=linear16"
            f"&sample_rate={sample_rate_hz}&channels=1&punctuate=true"
        )
        if not is_null_or_whitespace(language_hint):
            path += f"&language={_escape(language_hint)}"

        resp = await self._http.send_async(
            HttpRequest(
                method="POST",
                url=combine_uri(self._options.base_address, path),
                headers={"Authorization": f"Token {self._options.api_key or ''}"},
                body_bytes=bytes(audio_pcm16_mono),
                content_type="audio/raw",
            )
        )
        if not resp.is_success:
            self._logger.warning("Deepgram returned %s", resp.status_code)
            return _empty()

        doc = resp.json()
        if not isinstance(doc, dict):
            return _empty()

        # Response shape: results.channels[0].alternatives[0].transcript
        results = doc.get("results")
        if not isinstance(results, dict):
            return _empty()
        channels = results.get("channels")
        if not isinstance(channels, list) or len(channels) == 0:
            return _empty()
        first_channel = channels[0]
        if not isinstance(first_channel, dict):
            return _empty()
        alts = first_channel.get("alternatives")
        if not isinstance(alts, list) or len(alts) == 0:
            return _empty()
        first_alt = alts[0]
        if not isinstance(first_alt, dict):
            return _empty()

        text = first_alt.get("transcript") if isinstance(first_alt.get("transcript"), str) else ""

        segments: List[TranscribedSegment] = []
        words = first_alt.get("words")
        if isinstance(words, list):
            for w in words:
                if not isinstance(w, dict):
                    continue
                start = float(w["start"]) if isinstance(w.get("start"), (int, float)) else 0.0
                end = float(w["end"]) if isinstance(w.get("end"), (int, float)) else start
                confidence = (
                    float(w["confidence"]) if isinstance(w.get("confidence"), (int, float)) else 0.0
                )
                segments.append(
                    TranscribedSegment(
                        text=w.get("word") if isinstance(w.get("word"), str) else "",
                        offset=timedelta(seconds=start),
                        duration=timedelta(seconds=end - start),
                        language=language_hint,
                        confidence=confidence,
                    )
                )

        duration = timedelta(0)
        meta = doc.get("metadata")
        if isinstance(meta, dict) and isinstance(meta.get("duration"), (int, float)):
            duration = timedelta(seconds=float(meta["duration"]))

        return TranscriptionResult(text, language_hint, tuple(segments), duration)


__all__ = ["DeepgramSpeechRecognizer"]
